KAFKA-5629: Changed behavior of ConsoleConsumer when auto.offset.reset parameter is specified on the command line

soenkeliebau opened the pull request:
when "auto.offset.reset" property is specified on the command line but overridden by the code during startup. Currently the ConsoleConsumer silently overrides that setting, which can create confusing behavior.

soenkeliebau committed:
KAFKA-5629: Changed ConsoleConsumer to not override the value for auto.offset.reset with default, but honor parameter given in consumer.properties.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6305/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6289/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji commented:
Thanks for the patch. I guess another option is to accept the config value if no explicit `--from-beginning` is provided? Seems like the precedence could be:

1. Explicit --from-beginning
2. Config value
3. Default value ("latest")

We might also throw an error if the config value conflicts with an explicit `--from-beginning`.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6454/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6439/
Test PASSed (JDK 8 and Scala 2.12).

soenkeliebau commented:
I'm happy to change the PR to accommodate the behavior you detailed. I was a bit hesitant to change the implementation of a public facing cli without prior discussion. It would make more sense to implement latest as default but honor any given value.

hachikuji commented:
@soenkeliebau Yeah, that makes sense. I think in this case it's reasonable to change the behavior since 1) it's closer to how most users would expect the tool to work and 2) it seems unlikely that users depend on the default value overriding an explicitly passed configuration. We could mention the change in the release notes, but it might be overkill.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6600/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6601/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6605/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6616/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6620/
Test PASSed (JDK 7 and Scala 2.11).

soenkeliebau commented:
I've updated the pull request to behave like discussed. The JDK8 test currently fails due to a disc on the build server being full, I'll request a retest tomorrow but don't really expect any failures tbh.
I've added a few tests, skipping the error case, as I was unsure how to properly test for that. My first try caused the test to hang on the build server.

soenkeliebau commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6631/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6616/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji commented on core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:
```diff
@@ -176,6 +177,89 @@ class ConsoleConsumerTest {
+    assertFalse(config.useOldConsumer)
+    assertEquals("localhost:9092", config.bootstrapServer)
+    assertEquals("test", config.topicArg)
+    //assertEquals("earliest", config.extraConsumerProps.getProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG))
```
nit: Should we remove this line? There's a couple more in the test cases below.

soenkeliebau replied:
Absolutely, missed removing that when copying the tests. I'll delete these.

hachikuji commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -196,14 +196,31 @@ object ConsoleConsumer extends Logging {
     props
   }
 
+  def setAutoOffsetResetValue(config: ConsumerConfig, props: Properties) {
```
Maybe worth a comment mentioning that this handles the old and new consumers which share the same config property?

soenkeliebau replied:
Will add.

hachikuji commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -196,14 +196,31 @@ object ConsoleConsumer extends Logging {
+      if (config.options.has(config.resetBeginningOpt) && "latest".equals(props.getProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG))) {
+        // conflicting options - latest und earliest, throw an error
+        System.err.println("Can't simultaneously specify --from-beginning and 'auto.offset.reset=latest', please remove one option")
+        Exit.exit(1)
```
I'm guessing we don't have a test case for this because of the system exit? However, it looks like the `Exit` object does allow us to override the behavior with `Exit.setExitProcedure`. Maybe we could use that in a test case? Just need to be careful to reset the state with `resetExitProcedure`.

soenkeliebau replied:
That is right. I experimented with the solution from [this SO answer](https://stackoverflow.com/a/15994622/7183757) for a bit and it worked for me locally, but caused the build on jenkins to hang in that test case for a long time, probably because it wasn't allowed to shutdown. So I figured I'd leave it for now, but I will give your approach a try.

hachikuji reviewed (commented):
Thanks for the update. Looks good overall. Left a few minor comments.

soenkeliebau commented:
I've made some changes to the PR as per your comments, added a comment, removed the lines and created a test case to check that the console consumer correctly fails on conflicting parameter values.

The tests don't seem to run currently, I'll wait until tomorrow if this is just a small delay and request a retest then, but don't expect anything to fail, as I didn't change the code and the new test case worked well locally (but you never know).

hachikuji reviewed (approved):
LGTM. Thanks for the patch. I'll merge after the tests finish.

soenkeliebau commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6667/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6652/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji commented:
@soenkeliebau Merged to trunk. I modified the title when merging. Hope you don't mind.

soenkeliebau commented:
Absolutely not, thanks for your help!
